src/VisualStudio/Core/Def/Implementation/UnusedReferences/RemoveUnusedReferencesCommandHandler.cs

```diff
@@ -0,0 +1,319 @@
+using Microsoft.CodeAnalysis.UnusedReferences;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.LanguageServices.Implementation.UnusedReferences.Dialog;
+using Microsoft.VisualStudio.LanguageServices.Implementation.UnusedReferences.ProjectAssets;
+using Microsoft.VisualStudio.PlatformUI;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.UnusedReferences
+{
+    [Export(typeof(RemoveUnusedReferencesCommandHandler)), Shared]
+    internal sealed class RemoveUnusedReferencesCommandHandler
+    {
+        private const string ProjectAssetsFilePropertyName = "ProjectAssetsFile";
+
+        private readonly IReferenceCleanupService _referenceCleanupService;
+        private readonly IUnusedReferencesService _unusedReferencesService;
+        private readonly RemoveUnusedReferencesDialogProvider _unusedReferenceDialogProvider;
+        private readonly VisualStudioWorkspaceImpl _workspace;
+        private readonly IVsHierarchyItemManager _vsHierarchyItemManager;
+        private readonly IUIThreadOperationExecutor _threadOperationExecutor;
+        private IServiceProvider? _serviceProvider;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public RemoveUnusedReferencesCommandHandler(
+            RemoveUnusedReferencesDialogProvider unusedReferenceDialogProvider,
+            IVsHierarchyItemManager vsHierarchyItemManager,
+            IUIThreadOperationExecutor threadOperationExecutor,
+            VisualStudioWorkspaceImpl workspace)
+        {
+            _unusedReferenceDialogProvider = unusedReferenceDialogProvider;
+            _vsHierarchyItemManager = vsHierarchyItemManager;
+            _threadOperationExecutor = threadOperationExecutor;
+            _workspace = workspace;
+
+            _referenceCleanupService = workspace.Services.GetRequiredService<IReferenceCleanupService>();
+            _unusedReferencesService = workspace.Services.GetRequiredService<IUnusedReferencesService>();
+        }
+
+        public void Initialize(IServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider;
+
+            Contract.ThrowIfNull(_serviceProvider);
+
+            // Hook up the "Remove Unused References" menu command for CPS based managed projects.
+            var menuCommandService = (IMenuCommandService)_serviceProvider.GetService(typeof(IMenuCommandService));
+            if (menuCommandService != null)
+            {
+                AddCommand(menuCommandService, ID.RoslynCommands.RemoveUnusedReferences, Guids.RoslynGroupId, OnRemoveUnusedReferencesForSelectedProject, OnRemoveUnusedReferencesForSelectedProjectStatus);
+            }
+
+            return;
+
+            // Local functions
```

Comment:
Why the local function? It's only called in one place?

Reply:
Moved to a helpers class since it was also implemented in teh VisualStudioDiagnosticAnalyzerService